src/LuaTransformer.ts

```diff
@@ -3370,17 +3502,146 @@ export class LuaTransformer {
+            }
+        }
+
+        return false;
+    }
+
+    protected replaceStatementInParent(oldNode: tstl.Statement, newNode?: tstl.Statement): void {
+        if (!oldNode.parent) {
+            throw new Error("node has not yet been assigned a parent");
+        }
+
+        if (tstl.isBlock(oldNode.parent) || tstl.isDoStatement(oldNode.parent)) {
+            if (newNode) {
+                oldNode.parent.statements.splice(oldNode.parent.statements.indexOf(oldNode), 1, newNode);
+            } else {
+                oldNode.parent.statements.splice(oldNode.parent.statements.indexOf(oldNode), 1);
+            }
+        } else {
+            throw new Error("unexpected parent type");
+        }
+    }
+
+    protected performHoisting(statements: tstl.Statement[]): tstl.Statement[] {
+        if (this.options.noHoisting) {
+            return statements;
+        }
+
+        const scope = this.peekScope();
+        const result = statements.slice();
+
+        // Hoist function definitions
+        if (scope.functionDefinitions) {
+            const hoistedFunctions: tstl.AssignmentStatement[] = [];
+            for (const [functionSymbolId, functionDefinition] of scope.functionDefinitions) {
+                if (this.shouldHoist(functionSymbolId, scope)) {
+                    const i = result.indexOf(functionDefinition.assignment);
+                    result.splice(i, 1);
+                    hoistedFunctions.push(functionDefinition.assignment);
+                }
+            }
+            if (hoistedFunctions.length > 0) {
+                result.unshift(...hoistedFunctions);
+            }
+        }
+
+        // Hoist variable declarations
+        if (scope.variableDeclarations) {
+            const hoistedLocals: tstl.Identifier[] = [];
+            for (const declaration of scope.variableDeclarations) {
+                const symbols = declaration.left.map(i => i.symbolId).filter(s => s !== undefined);
+                if (symbols.some(s => this.shouldHoist(s, scope))) {
+                    let assignment: tstl.AssignmentStatement | undefined;
+                    if (declaration.right) {
+                        assignment = tstl.createAssignmentStatement(
+                            declaration.left,
+                            declaration.right
+                        );
+                    }
+                    const i = result.indexOf(declaration);
+                    if (i >= 0) {
+                        if (assignment) {
+                            result.splice(i, 1, assignment);
+                        } else {
+                            result.splice(i, 1);
+                        }
+                    } else {
+                        // Special case for 'var's declared in child scopes
+                        this.replaceStatementInParent(declaration, assignment);
+                    }
+                    hoistedLocals.push(...declaration.left);
+                }
+            }
```

Comment:
This for loop + 3592 would probably also be nice to refactor to its own function